### PR TITLE
JAMES-3643 Local part resolution with virtual hosting

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraLocalPartLookupRepository.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package org.apache.james.user.cassandra;
```
Comment (Arsnael) [inline server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraLocalPartLookupRepository.java:1]: Missing license

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/LocalPartLookupModule.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package org.apache.james.user.cassandra;
```
Comment (Arsnael) [inline server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/LocalPartLookupModule.java:1]: Missing license

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/SessionProviderImpl.java [hunk 3]
```diff
@@ -59,16 +62,14 @@ public MailboxSession createSystemSession(Username userName) {
...
-        }
+        return validLogin(userid, passwd)
+            .map(Throwing.function(username -> createSession(username, MailboxSession.SessionType.User)))
+            .orElseThrow(BadCredentialsException::new);
     }
 
     @Override
     public MailboxSession loginAsOtherUser(Username adminUserid, String passwd, Username otherUserId) throws MailboxException {
-        if (! isValidLogin(adminUserid, passwd)) {
+        if (! validLogin(adminUserid, passwd).map(adminUserid::equals).orElse(false)) {
```
Comment (Arsnael) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/SessionProviderImpl.java:72]:
```suggestion
        if (!validLogin(adminUserid, passwd).map(adminUserid::equals).orElse(false)) {
```

File: server/data/data-jpa/src/main/java/org/apache/james/user/jpa/JPAUsersDAO.java [hunk 4]
```diff
@@ -103,6 +104,28 @@ public void init() {
...
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+
+        try {
+            List<JPAUser> results = entityManager
+                .createNamedQuery("findUserByLocalPart", JPAUser.class)
+                .setParameter("name", localPart.asString() + "@%")
+                .getResultList();
+            return results.stream()
+                .map(JPAUser::getUserName)
+                .collect(ImmutableList.toImmutableList());
```
Comment (Arsnael) [inline server/data/data-jpa/src/main/java/org/apache/james/user/jpa/JPAUsersDAO.java:118]: Can't do it in one shot? Like:

```
            return entityManager
                .createNamedQuery("findUserByLocalPart", JPAUser.class)
                .setParameter("name", localPart.asString() + "@%")
                .getResultList()
                .stream()
                .map(JPAUser::getUserName)
                .collect(ImmutableList.toImmutableList());
```

Comment (chibenwa) [reply server/data/data-jpa/src/main/java/org/apache/james/user/jpa/JPAUsersDAO.java:118]: Yes we could

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
There seems to be little interest in the community, this work is out of date. Abort!

**Pull request closed without merging**